### PR TITLE
US352932 create new Pan Validator and Length Validator class

### PR DESCRIPTION
**Pull request by charliewhight**

### CODE REVIEW

File: AccessCheckoutSDK/AccessCheckoutSDKTests/validation/LengthValidatorTests.swift [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    
+    func testShouldReturnTrueIfTextMatchesRegexAndValidLengthsIsEmpty() {
+        XCTAssertTrue(lengthValidator.validate(text: "123", againstValidationRule: AccessCardConfiguration.CardValidationRule(matcher: "^\\d*$", validLengths: [])))
+    }
+    
+    func testShouldReturnTrueIfTextMatchesRegexAndIsEqualToAValidLength() {
+        XCTAssertTrue(lengthValidator.validate(text: "4111111111111111", againstValidationRule: cardValidationRule))
+    }
+    
+    func testShouldReturnFalseIfTextMatchesRegexAndIsShorterThanAValidLength() {
```
Comment (ochalet-wp) [inline AccessCheckoutSDK/AccessCheckoutSDKTests/validation/LengthValidatorTests.swift:25]: `ShorterThanMinValidLength`

File: AccessCheckoutSDK/AccessCheckoutSDKTests/validation/card/PanValidatorTests.swift [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    
+    let baseDefaults = AccessCardConfiguration.CardDefaults.baseDefaults()
+    
+    func testShouldReturnFalseWithCardBrandWhenKnownPanIsShorterThanValidLength() {
+        let cardConfiguration = AccessCardConfiguration(defaults: baseDefaults, brands: [visaBrand])
+        let result = PanValidator(
+            cardConfiguration: cardConfiguration
+        ).validatePan(pan: "4111")
+        
+        XCTAssertFalse(result.0)
```
Comment (ochalet-wp) [inline AccessCheckoutSDK/AccessCheckoutSDKTests/validation/card/PanValidatorTests.swift:30]: result.0 / result.1 tp be changed to use the actual properties name

File: AccessCheckoutSDK/AccessCheckoutSDKTests/validation/card/PanValidatorTests.swift [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        pans: [12,13,14,15,16,17,18,19]
+    )
+    
+    let baseDefaults = AccessCardConfiguration.CardDefaults.baseDefaults()
+    
+    func testShouldReturnFalseWithCardBrandWhenKnownPanIsShorterThanValidLength() {
+        let cardConfiguration = AccessCardConfiguration(defaults: baseDefaults, brands: [visaBrand])
+        let result = PanValidator(
+            cardConfiguration: cardConfiguration
+        ).validatePan(pan: "4111")
```
Comment (ochalet-wp) [inline AccessCheckoutSDK/AccessCheckoutSDKTests/validation/card/PanValidatorTests.swift:28]: Can we separate the set up of the test from the action performed?

```
let validator = PanValidator(cardConfiguration: cardConfiguration)
let result = validator.validate(pan: "4111")
```

File: AccessCheckoutSDK/AccessCheckoutSDK/validation/card/AccessCardConfiguration.swift [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            return brands.first { $0.matcher.regexMatches(text: pan) == true }
+        }
+    }
+    
+    struct CardValidationRule: Decodable, Equatable {
+        var matcher: String?
+        var validLengths: Array<Int> = [Int]()
+    }
+    
+    public struct CardDefaults: Decodable {
```
[the remark this answers is not in the document]
Comment (charliewhight) [reply AccessCheckoutSDK/AccessCheckoutSDK/validation/card/AccessCardConfiguration.swift]: updated

File: AccessCheckoutSDK/AccessCheckoutSDK/validation/card/PanValidator.swift [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import Foundation
+
+struct PanValidator {
+    
+    private var lengthValidator:LengthValidator = LengthValidator()
+    private var cardConfiguration: AccessCardConfiguration
+    
+    public init(cardConfiguration: AccessCardConfiguration) {
```
[the remark this answers is not in the document]
Comment (charliewhight) [reply AccessCheckoutSDK/AccessCheckoutSDK/validation/card/PanValidator.swift]: updated

File: AccessCheckoutSDK/AccessCheckoutSDK/validation/card/PanValidator.swift [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        self.cardConfiguration = cardConfiguration
+    }
+      
+    
+    func validatePan(pan: PAN) -> (isValid: Bool, cardBrand: AccessCardConfiguration.CardBrand?) {
```
[the remark this answers is not in the document]
Comment (charliewhight) [reply AccessCheckoutSDK/AccessCheckoutSDK/validation/card/PanValidator.swift]: updated

File: AccessCheckoutSDK/AccessCheckoutSDKTests/validation/LengthValidatorTests.swift (hunk 1 above)
Comment (charliewhight) [reply AccessCheckoutSDK/AccessCheckoutSDKTests/validation/LengthValidatorTests.swift:25]: updated

File: AccessCheckoutSDK/AccessCheckoutSDKTests/validation/card/PanValidatorTests.swift (hunk 2 above)
Comment (charliewhight) [reply AccessCheckoutSDK/AccessCheckoutSDKTests/validation/card/PanValidatorTests.swift]: updated

File: AccessCheckoutSDK/AccessCheckoutSDKTests/validation/card/PanValidatorTests.swift (hunk 3 above)
Comment (charliewhight) [reply AccessCheckoutSDK/AccessCheckoutSDKTests/validation/card/PanValidatorTests.swift]: updated

**Pull request merged**